### PR TITLE
Updated node20 and dependencies

### LINKED ISSUE #19: Node 20 and update dependencies

### PR DESCRIPTION
**Pull request by jpolavar**
Closes #19

### CODE REVIEW

File: .github/workflows/check-published-scheduled.yml [hunk 1]
```diff
@@ -4,7 +4,7 @@ on:
   workflow_dispatch:
   schedule:
     - cron: '30 20 * * *'
-
+    - cron: '30 20 * * *'
```
Comment (le-cong) [inline .github/workflows/check-published-scheduled.yml:7]: duplicated?

File: .github/workflows/coverage.yml [hunk 2]
```diff
@@ -13,17 +13,18 @@ jobs:
     runs-on: ubuntu-latest
     name: Coverage
     steps:
-      - name: Checkout Code
-        uses: actions/checkout@v3
+      - uses: actions/checkout@v3
```
Comment (le-cong) [inline .github/workflows/coverage.yml:16]: i think we are on actions/checkout@v4 now, maybe checkout the latest public-library-template repo?

File: .github/workflows/publish.yml [hunk 3]
```diff
@@ -15,14 +15,18 @@ jobs:
         uses: actions/setup-node@v3
         with:
           node-version: '20.x'
-          registry-url: 'https://registry.npmjs.org'
+          cache: 'npm'
```
Comment (le-cong) [inline .github/workflows/publish.yml:18]: same here

File: package.json [hunk 4]
```diff
@@ -65,8 +69,8 @@
     "lint:fix": "eslint --ignore-path .gitignore . --fix",
     "test": "npm run ci:compile && npm run ci:test && npm run ci:lint && npm run ci:style",
     "ci:compile": "tsc --noEmit",
-    "ci:test": "jest --coverage=false",
-    "ci:coverage": "jest --coverage=true",
+    "ci:test": "NODE_OPTIONS=\"--no-warnings --experimental-vm-modules\" jest --coverage=false",
```
Comment (le-cong) [inline package.json:72]: remove --no-warnings?

File: package.json [hunk 5]
```diff
@@ -41,18 +41,22 @@
...
   },
   "eslintConfig": {
     "extends": [
       "@checkdigit/eslint-config"
     ]
   },
   "jest": {
-    "preset": "@checkdigit/jest-config"
+    "preset": "@checkdigit/jest-config",
+    "moduleFileExtensions": [
```
Comment (le-cong) [inline package.json:56]: is `moduleFileExtensions` necessary?

File: src/country.spec.ts [hunk 6]
```diff
@@ -51,16 +52,16 @@ describe('country', () => {
 
     assert.throws(
       () => country(at).getCountry(undefined as unknown as CountryAlpha2),
-      /^TypeError: Country not found for code 'undefined'$/u,
+      /^TypeError: Country not found for code 'undefined'$/u, // This is to check for the Country not found as undefined error
```
Comment (le-cong) [inline src/country.spec.ts:55]: this seems to be an exact match, so maybe just use plain string instead of regex?

File: src/format.ts [hunk 7]
```diff
@@ -51,7 +51,10 @@ export default function (at: string): FormatLibrary {
       let majorUnitAmount: number | bigint =
         Number(amount) === 0 ? Number(amount) / Number(minorUnit) : amountInteger / minorUnit;
 
-      if (!resolvedOptions.useDecimal && !(!resolvedOptions.useCurrency && !resolvedOptions.useGrouping)) {
+      if (
+        !resolvedOptions.useDecimal &&
+        !(!resolvedOptions.useCurrency && resolvedOptions.useGrouping !== undefined && !resolvedOptions.useGrouping)
```
Comment (le-cong) [inline src/format.ts:56]: maybe simply as:
```
if (
        !resolvedOptions.useDecimal &&
        !(!resolvedOptions.useCurrency && resolvedOptions.useGrouping === false)
      ) {
```

**Review by le-cong — CHANGES_REQUESTED** [review]
just a few comments

**Review by le-cong — CHANGES_REQUESTED** [review]
nothing special, just noticed some differences with the PSG standards (.yml, and package.json, etc.) can you double check?

**Comment by jpolavar** [comment]
> nothing special, just noticed some differences with the PSG standards (.yml, and package.json, etc.) can you double check?

We don't need prepare script  in package file so we need to change .yml files as well

File: src/format.ts [hunk 8]
```diff
@@ -51,7 +51,7 @@ export default function (at: string): FormatLibrary {
       let majorUnitAmount: number | bigint =
         Number(amount) === 0 ? Number(amount) / Number(minorUnit) : amountInteger / minorUnit;
 
-      if (!resolvedOptions.useDecimal && !(!resolvedOptions.useCurrency && !resolvedOptions.useGrouping)) {
+      if (!resolvedOptions.useDecimal && !(!resolvedOptions.useCurrency && resolvedOptions.useGrouping === false)) {
```
[the remark this answers is not in the document]
Comment (jpolavar) [reply src/format.ts:54]: I updated  earlier as   ```
!resolvedOptions.useDecimal &&
        !(!resolvedOptions.useCurrency && resolvedOptions.useGrouping !== undefined && !resolvedOptions.useGrouping) 
```based on previous PR comment https://github.com/checkdigit/currency/pull/20#discussion_r1471756343 , updated as       `if (!resolvedOptions.useDecimal && !(!resolvedOptions.useCurrency && resolvedOptions.useGrouping === false)) {
`

[the remark this answers is not in the document]
Comment (le-cong) [reply src/format.ts:54]: i think my previous comment is not 100% acurate and missed the condition branch of checking `undefined`. what do you think of this:
```
if (!resolvedOptions.useDecimal && !(!resolvedOptions.useCurrency && resolvedOptions.useGrouping !== true)) {
        throw new Error('useDecimal can only be false if useCurrency and useGrouping are also false');
      }
```

[the remark this answers is not in the document]
Comment (jpolavar) [reply src/format.ts:54]: I added earlier```
 if (!resolvedOptions.useDecimal && !(!resolvedOptions.useCurrency && resolvedOptions.useGrouping === false)) {
        throw new Error('useDecimal can only be false if useCurrency and useGrouping are also false');
      }
``` same as the above one?

**Review by le-cong — APPROVED** [review]
lgtm

**Review by le-cong — CHANGES_REQUESTED** [review]
package.json is missing "type":"module"
there are still mismatch in CI config files such as ci.yml/coverage.yml/etc.
some diff tools can help to compare folder contents between the service and `public-library-template` repo

**Comment by jpolavar** [comment]
> package.json is missing "type":"module" there are still mismatch in CI config files such as ci.yml/coverage.yml/etc. some diff tools can help to compare folder contents between the service and `public-library-template` repo

We don't have npm run prepare in our package script for currency library and so coverage and ci yaml's are different.

**Review by le-cong — CHANGES_REQUESTED** [review]
sorry, but i'm still not sure if it's compared with the google doc or the `public-library-template` repo, The Luke Document doesn't differenticate ci.yml/coverage.yml between services and libraries, that's why it's better to refer to the template repositories for comparison IMO.
image

**Comment by jpolavar** [comment]
> sorry, but i'm still not sure if it's compared with the google doc or the `public-library-template` repo, The Luke Document doesn't differenticate ci.yml/coverage.yml between services and libraries, that's why it's better to refer to the template repositories for comparison IMO. image

But latest doc shows that we are using `cache: 'npm' ` instead of `registry-url:` 
Screenshot 2024-02-14 at 10 52 59 PM

**Review by le-cong — CHANGES_REQUESTED** [review]
the stuff in Luke Document doesn't cover everything, the ci.yml/coverage.yml listed there are intended for service, they are not for public repo. public repo has different files because the npm permission etc requirements are different, and that's why please look at https://github.com/checkdigit/public-library-template instead.

**Comment by ramaghanta** [comment]
> the stuff in Luke Document doesn't cover everything, the ci.yml/coverage.yml listed there are intended for service, they are not for public repo. public repo has different files because the npm permission etc requirements are different, and that's why please look at https://github.com/checkdigit/public-library-template instead.

We should not have to look at multiple sources for updating dependencies and the latest standards. So, let's update the Luke document with what the workflows should be for a public repo.

**Review by le-cong — APPROVED** [review]
lgtm

**Pull request merged**